### PR TITLE
Honor CHERI_USE_CAP_{TABLE,TLS} for all purecap architectures.

### PR DESCRIPTION
**Pull request by bsdjhb**

### CODE REVIEW

**Comment by bsdjhb** [comment]
I'm naively assuming that both of these knobs are architecture independent and not MIPS-specific.  I know that the captable one is at least.

One possible change to make here is to not force a captable ABI of pcrel if the option isn't set, but instead use the compiler's default ABI unless an explicit ABI is given?  I think the default ABI is probably pcrel on RISC-V and MIPS already and that forcing this to pcrel was probably a holdover from transitioning away from the legacy ABI on MIPS?

**Comment by jrtc27** [comment]
Only MIPS has an implementation for the non-default options. There is no legacy TLS on RISC-V (and maybe also gone on MIPS now), and pcrel is the only RISC-V captable ABI.

**Comment by bsdjhb** [comment]
I believe we do intend to add the plt captable ABI for RISC-V at least?  I'm happy to make the TLS one MIPS-only.

**Comment by arichardson** [comment]
The TLS option can probably be GC'd since I've deleted the compiler support.

**Comment by jrtc27** [comment]
TLS one should just go. I wouldn't be surprised if it's bit-rotted, and we're not setting DDC to non-null for purecap binaries so it can't possibly work.

Yes, RISC-V will gain other captable ABIs too. I do think though we want to check `defined(CHERI_USE_CAP_TABLE)` and let undef be the compiler's default.

**Comment by arichardson** [comment]
Making the ABI available for all seems fine to me (even if MIPS is the only architecture where non-pcrel is supported). We may want to rename the flag eventually to remove the "captable" from the name.

**Comment by bsdjhb** [comment]
I've axed the TLS option entirely and made the captable one only override if set.  I've also pulled in a change I had from the purecap branch to pull the subobject bounds bits over from bsd.cheri.mk to be available on all purecap architectures.

File: share/mk/bsd.cheri.mk [hunk 1]
```diff
@@ -120,9 +120,6 @@ ROOTOBJDIR=	${OBJTOP}/obj-libcheri
 .ifdef CHERI_USE_CAP_TABLE
 CFLAGS+=	-cheri-cap-table-abi=${CHERI_USE_CAP_TABLE}
 .endif
```
Comment (jrtc27) [inline share/mk/bsd.cheri.mk:122]: This should go

File: share/mk/bsd.cpu.mk [hunk 2]
```diff
@@ -484,3 +480,23 @@ MACHINE_ABI+=	ptr64
 .else
 MACHINE_ABI+=	ptr32
 .endif
+
+.if ${MACHINE_ABI:Mpurecap}
+.ifdef CHERI_USE_CAP_TABLE
+FLAGS+=	-cheri-cap-table-abi=${CHERI_USE_CAP_TABLE}
```
Comment (jrtc27) [inline share/mk/bsd.cpu.mk:486]:
```suggestion
CFLAGS+=	-cheri-cap-table-abi=${CHERI_USE_CAP_TABLE}
```

Comment (bsdjhb) [reply share/mk/bsd.cpu.mk]: Yeah, I found that right after I pushed it. :(

File: share/mk/bsd.cheri.mk (hunk 1 above)
Comment (bsdjhb) [reply share/mk/bsd.cheri.mk:122]: Hmm, it's not clear to me if bsd.cheri.mk even gets used anymore (or rather, when it still gets used).  I think the mips64c128 MACHINE_ARCH doesn't use it but depends on bsd.cpu.mk only.

See this bit at the top of bsd.cheri.mk:

```
.if ${MACHINE_ARCH:Mmips*} && (!${MACHINE_ARCH:Mmips*c*} || defined(COMPAT_CHERI))
```

So only hybrid MIPS or the libcheri built along with hybrid MIPS uses bsd.cheri.mk I think?  I think the purecap MIPS world no longer uses this at all.  I wonder how far away we are from disabling it for the hybrid MIPS world as well.  Probably a @brooksdavis question.

**Comment by bsdjhb** [comment]
> My only concern now is that bsd.compat.mk doesn't see this?

Hmmm, I think for the libcheri stage, bsd.compat.mk should arrange for the MACHINE_ARCH, etc. to all be set on the command line of the make that invokes 'make libraries' so bsd.cpu.mk should be honored.  However, if a leaf Makefile uses NEED_COMPAT/WANT_COMPAT, I'm not sure bsd.cpu.mk is re-parsed after MACHINE_ARCH is overridden at the bottom of bsd.compat.mk.  I think that's a larger problem though than this change.

**Comment by bsdjhb** [comment]
Which is to say, I think that's already broken if you put `WANT_COMPAT=CHERI` in a leaf Makefile today and isn't made "more broken" by this change.

**Comment by jrtc27** [comment]
> > My only concern now is that bsd.compat.mk doesn't see this?
> 
> Hmmm, I think for the libcheri stage, bsd.compat.mk should arrange for the MACHINE_ARCH, etc. to all be set on the command line of the make that invokes 'make libraries' so bsd.cpu.mk should be honored. However, if a leaf Makefile uses NEED_COMPAT/WANT_COMPAT, I'm not sure bsd.cpu.mk is re-parsed after MACHINE_ARCH is overridden at the bottom of bsd.compat.mk. I think that's a larger problem though than this change.

Yeah, I was thinking about the `.include <bsd.compat.mk>` case outside of Makefile.libcompat. Though, the bsd.compat.mk "solution" is just to duplicate the native/compat-dependent logic, it's just another thing like -march (you _should_ also filter it _out_, but probably does nothing if left other than give "unused argument" warnings). A more satisfactory solution would be nice, but that's really an upstream issue.

**Comment by jrtc27** [comment]
> Which is to say, I think that's already broken if you put `WANT_COMPAT=CHERI` in a leaf Makefile today and isn't made "more broken" by this change.

For that, yes, but any NEED_CHERI=purecap consumers will regress when switched to NEED_COMPAT=CHERI and risk giving link-time errors when building purecap binaries in a hybrid world with a non-default captable ABI.

File: share/mk/bsd.cheri.mk (hunk 1 above)
Comment (brooksdavis) [reply share/mk/bsd.cheri.mk:122]: We're at the point where we should make a push to kill bsd.cheri.mk (probably after Thursday though).

**Comment by bsdjhb** [comment]
> > Which is to say, I think that's already broken if you put `WANT_COMPAT=CHERI` in a leaf Makefile today and isn't made "more broken" by this change.
> 
> For that, yes, but any NEED_CHERI=purecap consumers will regress when switched to NEED_COMPAT=CHERI and risk giving link-time errors when building purecap binaries in a hybrid world with a non-default captable ABI.

So for today I think this means duplicating some bits in bsd.compat.mk for now?

**Comment by jrtc27** [comment]
> > > Which is to say, I think that's already broken if you put `WANT_COMPAT=CHERI` in a leaf Makefile today and isn't made "more broken" by this change.
> > 
> > 
> > For that, yes, but any NEED_CHERI=purecap consumers will regress when switched to NEED_COMPAT=CHERI and risk giving link-time errors when building purecap binaries in a hybrid world with a non-default captable ABI.
> 
> So for today I think this means duplicating some bits in bsd.compat.mk for now?

Yep. Also, hmm, bsd.compat.mk just appends overrides to CFLAGS. I guess that means don't filter these out for freebsd64 and hope it doesn't warn too much?.. Though in theory a true hybrid ABI could be partially affected by these.

**Comment by bsdjhb** [comment]
Fixed a few more bsd.compat.mk things.  We might be close to ditching bsd.cheri.mk entirely after this now.  At the very least, I think there's some other cruft in there we can likely purge now such as -mstack-alignment.

**Pull request merged**